python/lsst/analysis/tools/tasks/makeMetricTable.py

```diff
@@ -28,81 +28,94 @@
+    dimensions=(),
+    defaultTemplates={"metricBundleName": ""},
 ):
     data = ct.Input(
         doc="Metric bundle to read from the butler",
         name="{metricBundleName}",
         storageClass="MetricMeasurementBundle",
         deferLoad=True,
-        dimensions=(
-            "skymap",
-            "tract",
-        ),
+        dimensions=(),
         multiple=True,
     )
 
-    skymap = ct.Input(
-        doc="The skymap that covers the tract that the data is from.",
-        name=BaseSkyMap.SKYMAP_DATASET_TYPE_NAME,
-        storageClass="SkyMap",
-        dimensions=("skymap",),
-    )
-
     metricTable = ct.Output(
         doc="A summary table of all metrics by tract.",
         name="{metricBundleName}Table",
         storageClass="ArrowAstropy",
-        dimensions=("skymap",),
+        dimensions=(),
     )
 
     def __init__(self, *, config=None):
         super().__init__(config=config)
-        if config.loadMatchedVisitData:
-            self.data = ct.Input(
-                doc="Metric bundle to read from the butler.",
-                name=config.connections.metricBundleName,
-                storageClass="MetricMeasurementBundle",
-                deferLoad=True,
-                dimensions=(
-                    "instrument",
-                    "skymap",
-                    "tract",
-                ),
-                multiple=True,
+        self.data = ct.Input(
+            doc=self.data.doc,
+            name=self.data.name,
+            storageClass=self.data.storageClass,
+            deferLoad=self.data.deferLoad,
+            dimensions=frozenset(sorted(config.inputDataDimensions)),
+            multiple=True,
+        )
+        self.metricTable = ct.Output(
+            doc=self.metricTable.doc,
+            name=self.metricTable.name,
+            storageClass=self.metricTable.storageClass,
+            dimensions=frozenset(sorted(config.outputDataDimensions)),
+        )
+        if "tract" in config.inputDataDimensions:
+            self.skymap = ct.Input(
+                doc="The skymap that covers the tract that the data is from.",
+                name=BaseSkyMap.SKYMAP_DATASET_TYPE_NAME,
+                storageClass="SkyMap",
+                dimensions=("skymap",),
             )
+        self.dimensions.update(frozenset(sorted(config.outputDataDimensions)))
 
 
-class MakeMetricTableConfig(pipeBase.PipelineTaskConfig, pipelineConnections=MakeMetricTableConnections):
-    loadMatchedVisitData = Field[bool](
-        doc="Changes the input connection when true. Necessary because of dimension mismatches.",
-        default=False,
+class MakeMetricTableConfig(
+    pipeBase.PipelineTaskConfig,
+    pipelineConnections=MakeMetricTableConnections,
+):
+    inputDataDimensions = ListField(
+        doc="Dimensions of the input data.",
+        default=(),
+        dtype=str,
+        optional=False,
+    )
+    outputDataDimensions = ListField(
```

Comment:
Good point (on the renaming). I've renamed `outputTableDimensions` as I feel that more closely described the actual output. Technically it's a metric table, rather than a metric or metric bundle.

Reply:
One the "would it ever...?" - the outputTableDimensions are specified at runtime and can take any valid dimension. Of course, after the datasetType has been created, then it's fixed in the butler repo, but that's always the case.